python/dgl/batched_graph.py

```diff
@@ -751,20 +758,159 @@ def _max_on(graph, typestr, feat):
+
+    Returns
+    -------
+    tensor
+        The obtained tensor.
+    """
+    data_attr, batch_num_objs_attr, _ = READOUT_ON_ATTRS[typestr]
+    data = getattr(graph, data_attr)
+    feat = data[feat]
+
+    if isinstance(graph, BatchedDGLGraph):
+        batch_num_objs = getattr(graph, batch_num_objs_attr)
+        max_n_objs = max(batch_num_objs)
+        index = []
+        for i, num_obj in enumerate(batch_num_objs):
+            index.extend(range(i * max_n_objs, i * max_n_objs + num_obj))
+        index = F.tensor(index)
+        dtype = F.dtype(feat)
+        ctx = F.context(feat)
+        feat_ = F.zeros((len(batch_num_objs) * max_n_objs, *F.shape(feat)[1:]), dtype, ctx) - float('inf')
+        feat_ = F.scatter_row(feat_, index, feat)
+        feat_ = F.reshape(feat_, (len(batch_num_objs), max_n_objs, *F.shape(feat)[1:]))
+        feat_ = F.softmax(feat_, 1)
+        feat_ = F.reshape(feat_, (len(batch_num_objs) * max_n_objs, *F.shape(feat)[1:]))
+        return F.gather_row(feat_, index)
+    else:
+        return F.softmax(feat, 0)
+
+def _broadcast_on(graph, typestr, feat_data):
+    """Internal function of broadcasting features to node or edge features
+    of a given field.
+
+    Parameters
+    ----------
+    graph : DGLGraph
+        The graph
+    typestr : str
+        'nodes' or 'edges'
+    feat_data : tensor
+        The feature to broadcast. Tensor shape is :math:`(*)` for single graph,
+        and :math:`(B, *)` for batched graph.
+
+    Returns
+    -------
+    tensor
+        The node/edge features tensor with shape :math:`(N, *)`.
+    """
+    _, batch_num_objs_attr, num_objs_attr = READOUT_ON_ATTRS[typestr]
+
+    if isinstance(graph, BatchedDGLGraph):
+        batch_num_objs = getattr(graph, batch_num_objs_attr)
+        index = []
+        for i, num_obj in enumerate(batch_num_objs):
+            index.extend([i] * num_obj)
+        index = F.tensor(index)
+        return F.gather_row(feat_data, index)
+    else:
+        n_objs = getattr(graph, num_objs_attr)()
+        return F.stack([feat_data] * n_objs, 0)
+
+def _topk_on(graph, typestr, feat, k, descending=True, idx=-1):
+    """Internal function to take graph-wise top-k features along the
+    given dimension of a given field.
+    If :attr:`reverse` is set to True, return the k smallest elements
+    instead.
+
+    Parameters
+    ---------
+    graph : DGLGraph
+        The graph
+    typestr : str
+        'nodes' or 'edges'
+    feat : str
+        The feature field name.
+    k : int
+        The k in "top-k".
+    descending : bool
+        Controls whether to return the largest or smallest elements.
+    idx : int
+        The key index we sort features along.
+
+    Returns
+    -------
+    Tensor:
+        Top-k features of the given graph with shape :math:`(K, D)`,
+        if the input graph is a BatchedDGLGraph a tensor with shape
+        :math:`(B, K, D)` we be returned, where math`B` is the batch
+        size.
```

Comment:
Should we also return the index of the top-k features?

Reply:
Some graph may have less than k nodes, the behavior of the index for top-k  features is not defined.

Reply:
DONE